crates/dbs-miniball/docs/MemoryVirtualization.md

```diff
@@ -0,0 +1,378 @@
+# 内存虚拟化
+
+## 概述
+内存虚拟化包含对虚拟机地址空间的管理和对 MMIO 资源的管理，通过 vm-memory 和 dbs-address-space crate 完成上述工作。除此之外，本文在最后介绍了加载内核到 Guest 内存的过程。
+
+## 内存管理
+在配置 KVM 后，开始初始化 guest memory。通过 `vm-memery`、`dbs-address-space` 完成。
+
+1. vm-memory：在典型的虚拟机监视器 (VMM) 中，有几个组件需要访问 VM 物理内存，例如引导加载程序、虚拟设备驱动程序、virtio 后端驱动程序和虚拟主机驱动程序。 vm-memory crate 提供了一组特征来将 VM 内存消费者与 VM 内存提供者分离。基于这些特征，VM 内存消费者可以在不知道 VM 内存提供者的实现细节的情况下访问 VM 的物理内存。
+1. dbs-address-space：虚拟机的地址空间管理器，它管理驻留在 Guest 物理地址空间中的内存和 MMIO 资源。
+
+### 虚拟内存地址转换
+内存虚拟化解决虚拟机里面的进程如何访问物理机上的内存的问题。MMU 中增加了一级页表（X86架构下叫EPT），MMU 通过该页表完成虚拟物理地址 (Guest Physical Address, GPA) 到主机物理地址 (Host Physical Address, HPA) 的转换。
+
+- GuestOS 本身有虚拟地址空间，用 GVA 表示。
+- 虚拟机认为自己独占整个内存空间，用 GPA 表示。
+- HostOS 本身有虚拟机地址空间，用 HVA 表示。
+- 宿主机本身有物理内存空间，用 HPA 表示。
+
+当 CPU 处于 Guest 模式时，CPU 发出的内存访问请求，MMU 需要经过两层转换：首先是完成由虚拟地址到虚拟机物理地址的转换，接着是完成由虚拟机物理地址到真实物理地址的转换。内存虚拟化的问题变成了 GVA -> HPA 的映射问题。GVA -> GPA 通过 GuestOS 页表映射。HVA -> HPA 通过 HostOS 页表映射。因此，只要建立 GPA -> HVA 的映射关系，即可解决内存虚拟化的问题。可以保证不同虚拟机在内存访问上的隔离性。
+
+在Linux这种使用虚拟地址的OS中，虚拟地址经过page table转换可得到物理地址。
+
+![address1.png](img/address1.png)
+> From [https://zhuanlan.zhihu.com/p/69828213](https://zhuanlan.zhihu.com/p/69828213)
+
+如果这个操作系统是运行在虚拟机上的，那么这只是一个中间的物理地址（Intermediate Phyical Address - IPA），需要经过VMM/hypervisor的转换，才能得到最终的物理地址（Host Phyical Address - **HPA**）。从VMM的角度，guest VM中的虚拟地址就成了**GVA**(Guest Virtual Address)，IPA就成了**GPA**(Guest Phyical Address)。
+
+![address2.png](img/address2.png)
+> From [https://zhuanlan.zhihu.com/p/69828213](https://zhuanlan.zhihu.com/p/69828213)
+
+
+可见，如果使用VMM，并且guest VM中的程序使用虚拟地址，那么就需要两次地址转换。
+
+![address3.png](img/address3.png)
+> From [https://zhuanlan.zhihu.com/p/69828213](https://zhuanlan.zhihu.com/p/69828213)
+
```

Comment:
Could you add a pic here to demonstrate traditional guest memory layout?

Reply:
added